texastribune/dot
# Retool all build systems and app structure

#### What's this PR do?
Pretty much starts over this app. Specifically:
- Deletes a ton of legacy files
- Creates new Webpack build system for the Vue-powered dashboard front end
- Creates a TypeScript configuration + compile process that works for both browser code and Node
- Creates a basic Express app w/ routing structure, API versioning, error views, static-file serving
- Configures Sentry error logging
- Configures ESLint and Prettier

#### Why are we doing this? How does it help us?
The beginning of a big ol' revamp of this app!

#### How should this be manually tested?
Add an `env-docker`, then add the following:
```
API_HOST=http://localhost:3000
DB_HOST=dot-db
DB_USER=postgres
DB_PASSWORD=postgres
DB_NAME=dot
DB_PORT=5432
NODE_PORT=3000

ENABLE_SENTRY=true
SENTRY_DSN=https://ea2663c9f8084a8892df8fb5559af25e@o197244.ingest.sentry.io/1317454
SENTRY_ENVIRONMENT=<your-name>
```
Then:
- `make -f Makefile.dev`
- `npm start`
- Confirm Node starts up and that there's a message containing "to restart at any time." After a few seconds, a Webpack build should happen.
- Edit `server/app.ts` and save. Confirm the Node server automatically restarts. Wait for Webpack.
- Visit `localhost:3000`. Confirm it redirects to `/dashboard/` and that there's a box with two ellipses on the screen (this is a very basic Vue component).
- Visit `/dashboard` and confirm it redirects to `/dashboard/` (with a trailing slash).
- Visit `/api/v1/test/` and confirm you get JSON with something about foo and bar.
- Confirm `/robots.txt` exists and has `Disallow: /`.
- Put the following above line 51 in `server.ts`: `throw new Error('bleh');`. Wait for the rebuild.
- Visit the root route and confirm you get an error trace. Wait a few seconds and confirm a Sentry alert occurs, too.
- Kill the server and do `npm run build && npm run start:prod`. Visit a few routes from above and confirm they work (this is how the app will run in production).
- Kill the server again and do `npm run lint`. Confirm a linting task runs without error.
- BONUS IF ASHLEY IS REVIEWING: Open any `.ts` file and remove a semicolon from something. Confirm Prettier adds it back on save.

#### How should this change be communicated to end users?
NA.

#### Are there any smells or added technical debt to note?
On the road to so much less.

#### What are the relevant tickets?
https://3.basecamp.com/3098728/buckets/736178/todos/2575666513
https://3.basecamp.com/3098728/buckets/736178/todos/2545304594

#### Have you done the following, if applicable:

#### TODOs / next steps:
So much.
